lib/Runtime/Language/SourceTextModuleRecord.cpp

```diff
@@ -165,25 +217,245 @@ namespace Js
+                    // We are not rejecting ambiguous resolution at this time.
+                    if (starExportedNames != nullptr)
+                    {
+                        starExportedNames->Map([&](PropertyId propertyId) {
+                            if (propertyId != defaultPropertyId && !localNames->Has(propertyId))
+                            {
+                                localNames->Prepend(propertyId);
+                            }
+                        });
+                    }
+                }
+#if DBG
+                else
+                {
+                    AssertMsg(false, "dependent modules should have been initialized");
+                }
+#endif
+            });
+        }
+        exportedNames = localNames;
+        return localNames;
+    }
+
+    // return false when "ambiguous". 
+    // otherwise nullptr means "null" where we have circular reference/cannot resolve.
+    bool SourceTextModuleRecord::ResolveExport(PropertyId exportName, ResolveSet* resolveSet, ExportModuleRecordList* exportStarSet, ModuleNameRecord** exportRecord)
+    {
+        ArenaAllocator* allocator = EnsureTempAllocator();
+        if (resolvedExportMap == nullptr)
         {
-            resolveSet = (ResolutionDictionary*)AllocatorNew(ArenaAllocator, allocator, ResolutionDictionary, allocator);
+            resolvedExportMap = AllocatorNew(ArenaAllocator, allocator, ResolvedExportMap, allocator);
         }
+        if (resolvedExportMap->TryGetReference(exportName, exportRecord))
+        {
+            return true;
+        }
+        // TODO: use per-call/loop allocator?
         if (exportStarSet == nullptr)
         {
-            exportStarSet = (ResolveSet*)AllocatorNew(ArenaAllocator, allocator, ResolveSet, allocator);
+            exportStarSet = (ExportModuleRecordList*)AllocatorNew(ArenaAllocator, allocator, ExportModuleRecordList, allocator);
+        }
+        if (resolveSet == nullptr)
+        {
+            resolveSet = (ResolveSet*)AllocatorNew(ArenaAllocator, allocator, ResolveSet, allocator);
+        }
+
+        *exportRecord = nullptr;
+        bool hasCircularRef = false;
+        resolveSet->MapUntil([&](ModuleNameRecord moduleNameRecord) {
+            if (moduleNameRecord.module == this && moduleNameRecord.bindingName == exportName)
+            {
+                *exportRecord = nullptr;
+                hasCircularRef = true;
+                return true;
+            }
+            return false;
+        });
+        if (hasCircularRef)
+        {
+            Assert(*exportRecord == nullptr);
+            return true;
         }
-        Assert(false);
-        return false;
+        resolveSet->Prepend({ this, exportName });
+
+        if (localExportRecordList != nullptr)
+        {
+            PropertyId localNameId = Js::Constants::NoProperty;
+            localExportRecordList->MapUntil([&](ModuleExportEntry exportEntry) {
```

Comment:
Once every new resolution. I cache the resolved result.
I was thinking about creating additional dictionary for all different lists, but I'm leaving it like this for now before we have more real cases. It'll be a problem if there are a lot of exports from one module, and I hope the "interface", ie. the exported entries, are more concise.

---
In reply to: [55075711](https://github.com/Microsoft/ChakraCore/pull/432#discussion_r55075711) [](ancestors = 55075711)